c++/src/kj/array.h

```diff
@@ -316,9 +316,26 @@ class HeapArrayDisposer final: public ArrayDisposer {
 
   virtual void disposeImpl(void* firstElement, size_t elementSize, size_t elementCount,
                            size_t capacity, void (*destroyElement)(void*)) const override;
+};
 
+class SmallArrayDisposer: public ArrayDisposer {
+public:
+  template <typename T>
+  static T* allocate(void* firstElement, size_t count);
   template <typename T>
-  struct Allocate_;
+  static T* allocateUninitialized(void* firstElement, size_t count);
+
+  static const SmallArrayDisposer instance;
+
+private:
+  static void* allocateImpl(void* firstElement, size_t elementSize, size_t elementCount,
+                     size_t capacity, void (*constructElement)(void*),
```

Comment:
Hmm, I wonder if having to virtualize the element constructor/destructor even in the small-buffer case will prove to be a significant performance hit vs. KJ_STACK_ARRAY.

With KJ_STACK_ARRAY we are assuming the type always has a default constructor and that it's OK for us to construct more elements than we actually use. Construction is probably pretty cheap in most if not all cases (probably just zeroing), though I suppose destruction could be doing a lot of wasteful branches checking if things are still null...

Reply:
I'm not sure we actually need the full logic of `HeapArrayDisposer`. If we simply require that the type has a default constructor, then we can allocate the small array as a static-sized array with all elements constructed upfront (which is what KJ_STACK_ARRAY does). In this case construction and destruction are just handled by the compiler and we don't need a disposer at all.